File: LittleGameClient/LittleGame/ClientManger/ClientSocketManager.cs
using System;
using System.Net.Sockets;
using System.Net;
using System.Threading;
using LittleGame.State;
using System.Collections.Generic;

namespace LittleGame.Sever
{
    class ClientSocketManager
    {
        public ClientPlayingState state;

        private Socket clientSocket;
        private Thread recvThread;

        private List<string> receivedMessages;
        public List<string> ReceivedMessages { get => receivedMessages; }
        private Mutex recvMsgMutex;
        public void AddReceivedMessage(string message)
        {
            recvMsgMutex.WaitOne();
            receivedMessages.Add(message);
            recvMsgMutex.ReleaseMutex();
        }
        public string PopReceivedMessage(int index)
        {
            string result = null;
            recvMsgMutex.WaitOne();
            if (index >= 0 && index <= ReceivedMessages.Count)
            {
                result = receivedMessages[index];
                receivedMessages.RemoveAt(index);
            }
            recvMsgMutex.ReleaseMutex();
            return result;
        }


        private int playerId;
        public int PlayerId { get => playerId; }
        private int playerNum;
        public int PlayerNum { get => playerNum; }

        private bool connected;
        public bool Connected { get => connected; }
        private bool gameStart;
        public bool GameStart { get => gameStart; }

        

        public ClientSocketManager()
        {
            connected = false;
            gameStart = false;
            receivedMessages = new List<string>();
            recvMsgMutex = new Mutex();
        }

        private void waitingConnect()
        {
            Thread.Sleep(1000);
            if (!connected)
            {
                CloseConnection();
            }
        }

        public void StartConnect(string IP, int port)
        {
            clientSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            Thread waitThread = new Thread(waitingConnect);
            waitThread.IsBackground = true;
            waitThread.Start();
            try
            {
                clientSocket.Connect(IPAddress.Parse(IP), port);
                connected = true;
                waitThread.Abort();
                byte[] bytes = new byte[2048];
                int ret = clientSocket.Receive(bytes);
                if(ret <= 0)
                {
                    return;
                }
                string message = System.Text.Encoding.UTF8.GetString(bytes);
                message = message.Replace("\n", "");
                Console.WriteLine(message);
                string[] messages = message.Split(';');
                for (int i = 0; i < messages.Length; i++)
                {
                    string[] messageArgs = messages[i].Split(',');
                    if (messageArgs[0] == "Full")
                    {
                        connected = false;
                        try
                        {
                            clientSocket.Shutdown(SocketShutdown.Both);
                        }
                        finally
                        {
                            clientSocket = null;
                        }
                    }
                    else if (messageArgs[0] == "Id")
                    {
                        playerId = int.Parse(messageArgs[1]);
                        connected = true;
                        recvThread = new Thread(rcvMessage);
                        recvThread.IsBackground = true;
                        recvThread.Start();
                    }
                    else if (messageArgs[0].Equals("PlayerNum"))
                    {
                        playerNum = int.Parse(messageArgs[1]);
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
        }

        public void CloseConnection()
        {
            connected = false;
            gameStart = false;
            try
            {
                if (clientSocket != null)
                {
                    clientSocket.Close();
                    clientSocket = null;
                }
                if (recvThread != null)
                {
                    recvThread.Abort();
                    recvThread = null;
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
        }
        

        public void SendMessage(string message)
        {
            message += ";";
            if (Connected)
            {
                try
                {
                    clientSocket.Send(System.Text.Encoding.UTF8.GetBytes(message));
                    Thread.Sleep(5);
                }
                catch (Exception e)
                {
                    Console.Write(e.ToString());
                    CloseConnection();
                }
            }
        }

        private void rcvMessage()
        {
            receivedMessages.Clear();
            while (Connected)
            {
                try
                {
                    byte[] bytes = new byte[2048];
                    int ret = clientSocket.Receive(bytes);
                    if(ret <= 0)
                    {
                        CloseConnection();
                        break;
                    }
                    string message = System.Text.Encoding.UTF8.GetString(bytes);
                    message = message.Replace("\n", "");
                    Console.WriteLine(message);
                    string[] messages = message.Split(';');
                    for (int i = 0; i < messages.Length; i++)
                    {
                        string[] messageArgs = messages[i].Split(',');

                        if (messageArgs[0].Equals("Start"))
                        {
                            gameStart = true;
                        }
                        else if (messageArgs[0].Equals("Id"))
                        {
                            playerId = int.Parse(messageArgs[1]);
                        }
                        else if (messageArgs[0].Equals("PlayerNum"))
                        {
                            playerNum = int.Parse(messageArgs[1]);
                        }
                        else if (messageArgs[0].Equals("Move"))
                        {
                            AddReceivedMessage(messages[i]);
                        }
                        else if (messageArgs[0].Equals("Face"))
                        {
                            AddReceivedMessage(messages[i]);
                        }
                        else if (messageArgs[0].Equals("Attack"))
                        {
                            AddReceivedMessage(messages[i]);
                        }
                        else if (messageArgs[0].Equals("Reload"))
                        {
                            AddReceivedMessage(messages[i]);
                        }
                        else if (messageArgs[0].Equals("ReloadDone"))
                        {
                            AddReceivedMessage(messages[i]);
                        }
                        else if (messageArgs[0].Equals("Hitted"))
                        {
                            AddReceivedMessage(messages[i]);
                        }
                        else if (messageArgs[0].Equals("GameOver"))
                        {
                            gameStart = false;
                        }
                        Thread.Sleep(5);
                    }
                }
                catch (Exception e)
                {
                    Console.Write(e.ToString());
                    CloseConnection();
                }
            }
        }

    }
}

File: LittleGame_Android/app/src/main/java/com/example/littlegame/ClientSocketManager.java
package com.example.littlegame;

import android.util.Log;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.Socket;
import java.util.Vector;

public class ClientSocketManager {

    private Socket clientSocket;
    private BufferedReader in;
    private BufferedWriter out;
    private String severIp;
    private int port;
    public Thread recvThread, sendThread;
    public Vector<String> messagesToSend;
    public Vector<String> recivedMessages;

    private boolean connecting;
    public boolean getConnecting() { return connecting; }

    private int playerId;
    public int getPlayerId() { return playerId; }
    private int playerNum;
    public int getPlayerNum() { return playerNum; }

    private boolean connected;
    public boolean getConnected() { return connected; }
    private boolean gameStart;
    public boolean getGameStart() { return gameStart; }


    public ClientSocketManager()
    {
        connected = false;
        gameStart = false;
        connecting = false;
        clientSocket = null;
        messagesToSend = new Vector<String>();
        recivedMessages = new Vector<String>();
        playerId = playerNum = 0;
    }

    public void StartConnect(String ip, int port) {
        this.severIp = ip;
        this.port = port;
        if(!connected && !connecting) {
            recvThread = new Thread(connection);
            recvThread.start();
        }
    }

    private Runnable connection = new Runnable() {
        @Override
        public void run() {
            try {
                connecting = true;
                clientSocket = new Socket(severIp, port);
                connected = true;
                if(connected) {
                    in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
                    out = new BufferedWriter(new OutputStreamWriter(clientSocket.getOutputStream(), "BIG5"));
                    sendThread = new Thread(snedMessages);
                    sendThread.start();
                }
                connecting = false;
                recivedMessages.clear();
                while (connected){
                    String message = in.readLine();
                    if(message!=null) {
                        Log.d("Debug", "recv message = " + message);
                        String[] messages = message.split(";");
                        for (int i = 0; i < messages.length; i++)
                        {
                            String[] messageArgs = messages[i].split(",");
                            if (messageArgs[0].equals("Full"))
                            {
                                connected = false;
                                Disconnet();
                            }
                            else if (messageArgs[0].equals("Id"))
                            {
                                playerId = Integer.parseInt(messageArgs[1]);
                            }
                            else if (messageArgs[0].equals("PlayerNum"))
                            {
                                playerNum = Integer.parseInt(messageArgs[1]);
                            }
                            else if (messageArgs[0].equals("Start"))
                            {
                                gameStart = true;
                            }
                            else if (messageArgs[0].equals("Move"))
                            {
                                recivedMessages.add(messages[i]);
                            }
                            else if (messageArgs[0].equals("Face"))
                            {
                                recivedMessages.add(messages[i]);
                            }
                            else if (messageArgs[0].equals("Attack"))
                            {
                                recivedMessages.add(messages[i]);
                            }
                            else if (messageArgs[0].equals("Reload"))
                            {
                                recivedMessages.add(messages[i]);
                            }
                            else if (messageArgs[0].equals("ReloadDone"))
                            {
                                recivedMessages.add(messages[i]);
                            }
                            else if (messageArgs[0].equals("Hitted"))
                            {
                                recivedMessages.add(messages[i]);
                            }
                            else if (messageArgs[0].equals("GameOver"))
                            {
                                gameStart = false;
                            }
                            Thread.sleep(5);
                        }
                    }
                }
            }
            catch (Exception e){
                e.printStackTrace();
                Disconnet();
            }
        }
    };

    private Runnable snedMessages = new Runnable() {
        @Override
        public void run() {
            try {
                messagesToSend.clear();
                String message = "";
                while (connected){
                    if(messagesToSend.size() > 0){
                        message = messagesToSend.remove(0);
                        out.write(message + "\n");
                        out.flush();
                    }
                }
            }
            catch (Exception e){
                e.printStackTrace();
                Disconnet();
            }
        }
    };

    public void AddMessageToSend(String message) {
        if(message.charAt(message.length()-1) != ';') {
            message += ";";
        }
        messagesToSend.add(message);
    }

    public void Disconnet() {
        Log.d("Debug", "disconnect");
        connected = false;
        connecting = false;
        gameStart = false;
        try
        {
            if (clientSocket != null)
            {
                clientSocket.close();
                clientSocket = null;
            }
            if (recvThread != null)
            {
                recvThread.interrupt();
                recvThread = null;
            }
            if (sendThread != null)
            {
                sendThread.interrupt();
                sendThread = null;
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}
